File: Minstack.java
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this : NO


// Your code here along with comments explaining your approach

import java.util.Stack;

class MinStack {

    int min;
    Stack<Integer> myStack;

    public MinStack() {
        this.myStack = new Stack<>();
        this.min = Integer.MAX_VALUE;
    }

    public void push(int val) {
        if (val <= min) {
            myStack.push(min);
            min = val;
        }
        myStack.push(val);
    }

    public void pop() {
        if (min == myStack.pop()) {
            min = myStack.pop();
        }
    }

    public int top() {
        return myStack.peek();
    }

    public int getMin() {
        return min;
    }
}

File: minstack.py
# Time complexity is o(1) and space complexity is O(N)

class MyQueue(object):

    def __init__(self):
        self.s1 = []
        self.s2 = []
        

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        self.s1.append(x)

    def pop(self):
        """
        :rtype: int
        """
        while not self.s2:
            while self.s1:
                self.s2.append(self.s1.pop())
        return self.s2.pop()                

    def peek(self):
        """
        :rtype: int
        """
        while not self.s2:
            while self.s1:
                self.s2.append(self.s1.pop())
        return self.s2[-1]

    def empty(self):
        """
        :rtype: bool
        """
        return max(len(self.s1), len(self.s2)) == 0 